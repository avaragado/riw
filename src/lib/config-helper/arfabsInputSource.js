// @flow

import glob from 'glob';
import chain from 'ramda/src/chain';

import type { Glob } from '../../types';
import type { FilesFromConfig } from './';

const arfabsInputSource: FilesFromConfig = config => chain(
    (sGlob: Glob) => glob.sync(
        sGlob,
        {
            cwd: config.rootDir,
            absolute: true,
        },
    ),
    config.sourceDirs,
);

export default arfabsInputSource;
