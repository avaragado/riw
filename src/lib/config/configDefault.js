// @flow

const cfg: RIWConfig = {
    configFile: null, // overridden
    rootDir: '', // overridden
    defaultLocale: 'en-us',
    targetLocales: [],
    translationsDatabaseFile: 'src/locale/riw-db.json',
    sourceDirs: ['src/**/*.js'],
    collateDir: 'tmp/babel-plugin-react-intl',
    inputMode: 'source',
    translationsOutputFile: 'src/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'src/locale/TODO-untranslated.json',
};

export default cfg;
