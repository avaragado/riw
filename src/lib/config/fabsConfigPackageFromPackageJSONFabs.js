// @flow

import path from 'path';

export default (fabsPackageJSON: ?AbsolutePath): ?AbsolutePath => (
    fabsPackageJSON
        ? path.resolve(path.dirname(fabsPackageJSON), '.riw-config.js')
        : null
    );
