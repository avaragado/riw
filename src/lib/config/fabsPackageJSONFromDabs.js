// @flow

import findUp from 'find-up';

export default (cwd: AbsolutePath): ?AbsolutePath => findUp.sync('package.json', { cwd });

