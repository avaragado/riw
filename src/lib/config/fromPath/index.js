// @flow

import path from 'path';
import chalk from 'chalk';

import type { Path, AbsolutePath } from '../../../types';
import type { ConfigSparse, ConfigSparseWithSource } from '../../config';
import log from '../../log';

export default (pathConfig: Path): ?ConfigSparseWithSource => {
    const fabs: AbsolutePath = path.resolve(pathConfig);

    try {
        return {
            ...(require(fabs): ConfigSparse),
            configFile: fabs,
        };

    } catch (err) {
        log.error('File not found: ', chalk.bold(fabs));
        return null;
    }
};
