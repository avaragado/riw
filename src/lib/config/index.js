// @flow

import configDefault from './resolve/configDefault';
import fabsConfigPackageForCWD from './fromPackage/fabsConfigPackageForCWD';
import configFromPath from './fromPath';
import configFromPackage from './fromPackage';
import configResolve from './resolve';

export {
    configDefault,
    fabsConfigPackageForCWD,
    configFromPath,
    configFromPackage,
    configResolve,
};
