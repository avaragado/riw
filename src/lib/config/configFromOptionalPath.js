// @flow

import path from 'path';

import log from '../log';

import configDefault from './configDefault';
import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

const requireOrNull = (fabs: AbsolutePath): ?Object => {
    try {
        return require(fabs);

    } catch (err) {
        return null;
    }
};

const configFromPath = (pathConfig: Path): ?RIWConfig => {
    const fabs: AbsolutePath = path.resolve(pathConfig);
    const configOverride = requireOrNull(fabs);

    if (configOverride === null) {
        log.error('riw', 'File not found: %j.', fabs);
        return null;
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabs),
    };
};

const configFromPackage = (): ?RIWConfig => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsConfigPackage || !fabsPackageJSON) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    let configOverride: ?Object = requireOrNull(fabsConfigPackage);

    if (configOverride === null) {
        configOverride = requireOrNull(fabsPackageJSON);

        if (configOverride && configOverride.riw) {
            configOverride = configOverride.riw;
        } else {
            configOverride = null;
        }
    }

    if (configOverride === null) {
        log.warn('riw', 'No config located for the project. You probably want to add some configuration.');
        configOverride = {};
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabsConfigPackage),
    };

};

export default (pathConfig: ?Path): ?RIWConfig => (
    pathConfig
        ? configFromPath(pathConfig)
        : configFromPackage()
);
