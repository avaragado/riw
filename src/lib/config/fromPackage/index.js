// @flow

import log from '../../log';

import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

export default (): ?RIWConfigSparseWithSource => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsPackageJSON || !fabsConfigPackage) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    try {
        return {
            ...(require(fabsConfigPackage): RIWConfigSparse),
            configFile: fabsConfigPackage,
        };

    } catch (err) {
        // fabsPackageJSON exists as we hunted the fs for it above:
        // but it might not have the riw key.
        const config: RIWConfigSparse = require(fabsPackageJSON).riw;

        if (!config) {
            log.warn('riw', 'No app config found. You probably want to add some configuration.');
        }

        return {
            ...config,
            configFile: fabsPackageJSON,
        };
    }
};
