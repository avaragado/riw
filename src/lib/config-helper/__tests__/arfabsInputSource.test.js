// @flow

import path from 'path';

import mock from 'mock-fs';

import type { ConfigSparseWithSource } from '../../config';
import { configResolve } from '../../config';

import arfabsInputSource from '../arfabsInputSource';

type Tree = { [key: string]: string | Tree };

type Fixture = {
    name: string,
    in: Tree,
    cfg: ConfigSparseWithSource,
};

const fixtures: Fixture[] = [
    {
        name: '01 empty dir -> empty array',
        in: {}, // empty dir
        cfg: {
            rootDir: '/fixtures',
            sourceDirs: ['dir/**/*.js'],
        },
    },

    {
        name: '02 sourceDirs elsewhere -> empty array',
        in: {
            'a.js': '//',
            'b.js': '//',
        },
        cfg: {
            rootDir: '/fixtures',
            // sourceDirs default not under dir, so result is empty array
        },
    },

    {
        name: '03 flat files, sourceDirs rel',
        in: {
            'a.js': '//',
            'b.js': '//',
        },
        cfg: {
            rootDir: '/fixtures',
            sourceDirs: ['dir/**/*.js'],
        },
    },

    {
        name: '04 flat files, sourceDirs abs',
        in: {
            'a.js': '//',
            'b.js': '//',
        },
        cfg: {
            rootDir: '/foobar',
            sourceDirs: [path.resolve('/fixtures', 'dir', '**/*.js')],
        },
    },

    {
        name: '05 nested files, multiple sourceDirs rel',
        in: {
            a: {
                'aa.js': '//',
                'ab.js': '//',
            },
            b: {
                'ba.js': '//',
                'bb.js': '//',
                'bc': {
                    'bcd.js': '//',
                },
            },
        },
        cfg: {
            rootDir: '/fixtures',
            sourceDirs: ['dir/a/*.js', 'dir/b/*.js'],
        },
    },

    {
        name: '06 nested files, multiple sourceDirs rel to parent of configFile',
        in: {
            a: {
                'aa.js': '//',
                'ab.js': '//',
            },
            b: {
                'ba.js': '//',
                'bb.js': '//',
                'bc': {
                    'bcd.js': '//',
                },
            },
        },
        cfg: {
            configFile: '/fixtures/package.json',
            sourceDirs: ['dir/a/*.js', 'dir/b/*.js'],
        },
    },
];

describe('lib/config-helper/arfabsInputSource', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                '/fixtures': {
                    dir: fixture.in,
                },
            });

            const cfg = configResolve(fixture.cfg);

            const received = arfabsInputSource(cfg);

            mock.restore();

            expect(received).toMatchSnapshot();
        });
    });
});
