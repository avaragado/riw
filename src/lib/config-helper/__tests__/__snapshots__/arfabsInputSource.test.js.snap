// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/config-helper/arfabsInputSource 01 empty dir -> empty array 1`] = `Array []`;

exports[`lib/config-helper/arfabsInputSource 02 sourceDirs elsewhere -> empty array 1`] = `Array []`;

exports[`lib/config-helper/arfabsInputSource 03 flat files, sourceDirs rel 1`] = `
Array [
  "/fixtures/dir/a.js",
  "/fixtures/dir/b.js",
]
`;

exports[`lib/config-helper/arfabsInputSource 04 flat files, sourceDirs abs 1`] = `
Array [
  "/fixtures/dir/a.js",
  "/fixtures/dir/b.js",
]
`;

exports[`lib/config-helper/arfabsInputSource 05 nested files, multiple sourceDirs rel 1`] = `
Array [
  "/fixtures/dir/a/aa.js",
  "/fixtures/dir/a/ab.js",
  "/fixtures/dir/b/ba.js",
  "/fixtures/dir/b/bb.js",
]
`;

exports[`lib/config-helper/arfabsInputSource 06 nested files, multiple sourceDirs rel to parent of configFile 1`] = `
Array [
  "/fixtures/dir/a/aa.js",
  "/fixtures/dir/a/ab.js",
  "/fixtures/dir/b/ba.js",
  "/fixtures/dir/b/bb.js",
]
`;
